Add sentinel support to eredis

dhruvjain99 opened the pull request:
Closes #43

dhruvjain99 commented:
@zuiderkwast could you please review this PR?

zuiderkwast commented on src/eredis_client.erl:
```diff
@@ -378,6 +387,32 @@ connect(#state{host = Host,
+
+    case eredis_sentinel:get_master() of
+        {ok, Host, Port} ->
+            #state{socket_options = SocketOptions,
+                tls_options = TlsOptions,
+                connect_timeout = ConnectTimeout,
+                auth_cmd = AuthCmd,
+                database = Db } = State,
```
Please align under {.

zuiderkwast commented on src/eredis_client.erl:
```diff
@@ -378,6 +387,32 @@ connect(#state{host = Host,
+                connect_timeout = ConnectTimeout,
+                auth_cmd = AuthCmd,
+                database = Db } = State,
+            case connect(Host, Port, SocketOptions, TlsOptions,
+                ConnectTimeout, AuthCmd, Db) of
```
```suggestion
            case connect(Host, Port, SocketOptions, TlsOptions,
                         ConnectTimeout, AuthCmd, Db) of
```

zuiderkwast commented on src/eredis_client.erl:
```diff
@@ -593,4 +628,4 @@ get_auth_command(undefined, "") -> % legacy
     eredis:create_multibulk([<<"AUTH">>, Password]);
 get_auth_command(Username, Password) ->
-    eredis:create_multibulk([<<"AUTH">>, Username, Password]).
+    eredis:create_multibulk([<<"AUTH">>, Username, Password]).
```
Don't remove the newline at end of file.

zuiderkwast commented on src/eredis_client.erl:
```diff
@@ -231,10 +234,15 @@ handle_info({reconnect, _Reason}, State) ->
+terminate(_Reason, #state{socket = Socket, transport = Transport, sentinel = undefined}) ->
+    Transport:close(Socket);
+terminate(_Reason, #state{socket = undefined}) ->
+    eredis_sentinel:stop();
```
If the eredis_sentinel process is shared between many eredis_client processes, if it's stopped it affects the other eredis clients which are still running, right?

zuiderkwast commented on src/eredis_client.erl:
```diff
@@ -378,6 +387,32 @@ connect(#state{host = Host,
+                    %% an expired certificate with TLS 1.3) schedule a reconnect already
+                    %% so that we don't try to reconnect if an error is received before
+                    %% reconnect_sleep milliseconds has elapsed.
+                    {ok, schedule_reconnect(unknown, State#state{socket = Socket, host = Host, port = Port})};
```
No need to set host and port in the state. On reconnect, connect(State) will be called, where sentinel is queried for host and port again.

Btw, there is some code duplication for the connect/1 function clauses with and without sentinel. Is it possible to get rid of it? For example something like this pseudo code:

```Erlang
connect(#state{...., host = Host0, port = Port0, sentinel = Sentinel, ....}) ->
    {Host, Port} = case Sentinel of
                       undefined -> {Host0, Port0};
                       SentinelPid -> eredis_sentinel:get_master(SentinelPid)
                   end,
    case connect(Host, Port, ......) of
        {ok, Socket} -> ....
        ConnError -> ConnError
    end.
```

zuiderkwast commented on src/eredis_sentinel.erl:
```diff
@@ -0,0 +1,251 @@
+    {noreply, S#eredis_sentinel_state{conn_pid = undefined}};
+
+%% Ignore late exit messages
+handle_info({'EXIT', _Pid, _Reason}, S) ->
+    {noreply, S};
```
If the `eredis_sentinel` process is linked to an `eredis_client` process which started the sentinel process (start_link) and the `eredis_client` process crashes, the `eredis_sentinel` process survives and is leaked.

It is risky to use `trap_exit` without a supervisor. I don't think it should be done. A supervisor can handle a process which ignores the 'EXIT' message by using brutal kill after a timeout, so it is the safe way to do it, or don't use trap exit if it's not needed. Can monitor be used instead?

dhruvjain99 replied:
So there are two ways. 
Since monitor is a unidirectional link, we'll need to pass eredis_sentinel process Pid to sentinel connection process so that it can setup a monitor link to avoid leak when sentinel crashes.
Or maybe we can keep trap_exit and stop the process on receiving exit except when it is from sentinel connection process by simply changing line 127 to `{stop, normal, S}`.

zuiderkwast replied:
You're right about monitor. I think we can't use it.

Trapping exit and handling all exit signals with `{stop, Reason, S}` would probably work.

However, handling processes like this is a bit like reinventing a supervisor. During the Code BEAM conference I talked to some people in the OTP team. They say that you should always use supervisors when you use multiple processes and links. The supervisor module is more robust and more tested. Reinventing the supervisor should not be done, because there is a risk of bugs, now or in code added later.

But how can we design a supervision tree for this? (I don't know.... Maybe we would need to use an `application` behaviour and add all eredis_sentinel processes to the eredis application supervisor? But starting the application is only useful when sentinel is used. Normally, eredis is not started as a separate application, but as part of the user's supervision trees. Maybe we can let `eredis_sentinel` to be an application, so it is only started when sentinel is used?....)

I think we can use trap exit it for now. We can say we have *experimental support* for sentinel, and maybe later we can change to use supervisors. What do you think?

dhruvjain99 replied:
As a first step to add sentinel support, it seems fine to me as well.

zuiderkwast reviewed (changes requested):
Thanks for your contribution!

I have a few concerns regarding the registered process(es), maybe we need one sentinel process per master group? And about trap exit, see comments.

Regarding the style, we use Emacs erlang-mode indentation. (We should add a CI job to check this, but we haven't done it yet.)

dhruvjain99 commented:
@zuiderkwast Thank you so much for a detailed review. I have resolved most of the comments and replied to a couple of them.

zuiderkwast commented on src/eredis_client.erl:
```diff
@@ -231,10 +234,17 @@ handle_info({reconnect, _Reason}, State) ->
+terminate(_Reason, #state{socket = undefined, sentinel = SentinelOptions}) ->
+    MasterGroup = proplists:get_value(master_group, SentinelOptions, mymaster),
+    eredis_sentinel:stop(MasterGroup);
+terminate(_Reason, #state{socket = Socket, transport = Transport, sentinel = SentinelOptions}) ->
+    Transport:close(Socket),
+    MasterGroup = proplists:get_value(master_group, SentinelOptions, mymaster),
+    eredis_sentinel:stop(MasterGroup).
```
Here we stop the sentinel for this master group, even if it was not started by this eredis client process. It means this affects other eredis client processes.

Maybe it works but it is strange.

How about we skip this and let the eredis_sentinel process handle the 'EXIT' signal when the linked eredis_client exits?

zuiderkwast commented on src/eredis_sentinel.erl:
```diff
@@ -0,0 +1,244 @@
+%% with Reason. The return value is ignored.
+-spec(terminate(Reason :: (normal | shutdown | {shutdown, term()} | term()),
+                State :: #eredis_sentinel_state{}) -> term()).
+terminate(_Reason, _State = #eredis_sentinel_state{conn_pid = Pid}) ->
+    eredis:stop(Pid),
```
Do we need to handle the case `conn_pid = undefined` here?

zuiderkwast replied:
If we want to cover these cases with test cases, we can find the linked processes using `process_info(Pid, links)` and kill or call `eredis:stop` on one of them. Just an idea.

dhruvjain99 commented:
@zuiderkwast I am trying to resolve the failing tests. When running `make ct-tcp` locally, few tests are failing. One of them is t-connect/1 with the following logs.
```
Testing lib.eredis.eredis_tcp_SUITE: Starting test, 53 test cases
%%% eredis_tcp_SUITE: 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
eredis_tcp_SUITE:'-t_connect/1-fun-2-' failed on line 94
Reason: {assertMatch,[{module,...},{...}|...]}
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

Testing lib.eredis.eredis_tcp_SUITE: *** FAILED test case 1 of 53 ***

=CRASH REPORT==== 30-Jun-2022::00:08:18.061638 ===
  crasher:
    initial call: eredis_client:init/1
    pid: <0.565.0>
    registered_name: []
    exception exit: {connection_error,econnrefused}
      in function  gen_server:init_it/6 (gen_server.erl, line 401)
    ancestors: [<0.561.0>]
    message_queue_len: 0
    messages: []
    links: [<0.561.0>]
    dictionary: []
    trap_exit: false
    status: running
    heap_size: 376
    stack_size: 28
    reductions: 1325
  neighbours:
    neighbour:
      pid: <0.563.0>
      registered_name: []
      initial call: eredis_client:init/1
      current_function: {inet_gethost_native,getit,2}
      ancestors: [<0.561.0>]
      message_queue_len: 0
      links: [<0.561.0>]
      trap_exit: false
      status: waiting
      heap_size: 233
      stack_size: 55
      reductions: 285
      current_stacktrace: [{inet_gethost_native,getit,2,
                      [{file,"inet_gethost_native.erl"},{line,490}]},
                  {inet,gethostbyname_tm_native,4,
                      [{file,"inet.erl"},{line,1295}]},
                  {inet,getaddrs_tm,3,[{file,"inet.erl"},{line,1246}]},
                  {inet,getaddrs,3,[{file,"inet.erl"},{line,679}]},
                  {eredis_client,get_addrs,1,
                      [{file,
                           "/Users/dhruvjain/Desktop/workspace/eredis/src/eredis_client.erl"},
                       {line,469}]},
                  {eredis_client,connect,7,
                      [{file,
                           "/Users/dhruvjain/Desktop/workspace/eredis/src/eredis_client.erl"},
                       {line,394}]},
                  {eredis_client,connect,1,
                      [{file,
                           "/Users/dhruvjain/Desktop/workspace/eredis/src/eredis_client.erl"},
                       {line,371}]},
                  {eredis_client,handle_info,2,
                      [{file,
                           "/Users/dhruvjain/Desktop/workspace/eredis/src/eredis_client.erl"},
                       {line,217}]}]
    neighbour:
      pid: <0.564.0>
      registered_name: []
      initial call: eredis_client:init/1
      current_function: {gen_server,loop,7}
      ancestors: [<0.561.0>]
      message_queue_len: 0
      links: [<0.561.0>]
      trap_exit: false
      status: waiting
      heap_size: 233
      stack_size: 12
      reductions: 1211
      current_stacktrace: [{gen_server,loop,7,[{file,"gen_server.erl"},{line,437}]},
                  {proc_lib,init_p_do_apply,3,
                            [{file,"proc_lib.erl"},{line,226}]}]
    neighbour:
      pid: <0.562.0>
      registered_name: []
      initial_call: {test_server_sup,timetrap,4}
      current_function: {test_server_sup,timetrap,4}
      ancestors: []
      message_queue_len: 0
      links: [<0.561.0>]
      trap_exit: false
      status: waiting
      heap_size: 233
      stack_size: 6
      reductions: 12
      current_stacktrace: [{test_server_sup,timetrap,4,
                                   [{file,"test_server_sup.erl"},{line,63}]}]
    neighbour:
      pid: <0.561.0>
      registered_name: []
      initial_call: {erlang,apply,2}
      current_function: {ct_util,call,2}
      ancestors: []
      message_queue_len: 0
      links: [<0.562.0>,<0.564.0>,<0.565.0>,<0.563.0>,<0.541.0>]
      trap_exit: false
      status: running
      heap_size: 4185
      stack_size: 51
      reductions: 5120
      current_stacktrace: [{ct_util,call,2,[{file,"ct_util.erl"},{line,1061}]},
                  {ct_framework,end_tc,6,
                                [{file,"ct_framework.erl"},{line,741}]},
                  {test_server_sup,framework_call,4,
                                   [{file,"test_server_sup.erl"},{line,778}]},
                  {test_server,do_end_tc_call1,4,
                               [{file,"test_server.erl"},{line,1430}]},
                  {test_server,run_test_case_eval1,6,
                               [{file,"test_server.erl"},{line,1360}]},
                  {test_server,run_test_case_eval,9,
                               [{file,"test_server.erl"},{line,1225}]}]

%%% eredis_tcp_SUITE ==> t_connect: FAILED
%%% eredis_tcp_SUITE ==> 
Failure/Error: ?assertMatch({ ok , _ }, eredis : start_link ( "127.0.0.1" , ? PORT , [ { reconnect_sleep , no_reconnect } ] ))
  expected: = { ok , _ }
       got: {error,{connection_error,econnrefused}}
      line: 94
```
Any heads up on solving this issue for local.
When I use my local redis-server and `rebar3 ct --suite=test/eredis_tcp_SUITE.erl`, all the tests pass

dhruvjain99 commented:
@zuiderkwast I have figured out the issue on my local. On mac `--net=host` option is not supported due to which `make test` was not working. I have added the step in makefile to start sentinel for the sentinel related tests due to which the github ci was failing.

zuiderkwast commented on priv/configs/redis_sentinel.conf:
```diff
@@ -0,0 +1,10 @@
+daemonize yes
+sentinel myid 10022d3099331ec79bd29f9179c82be7bec80441
+pidfile "/var/run/redis.pid"
+user default on nopass ~* +@all
```
The CI job seems to fail on this line for Redis 4-5. I don't think there was ACL in these Redis versions.

dhruvjain99 commented:
Can we please rerun the CI jobs? Install elvis step failed.
```
Cloning into 'elvis'...
Note: switching to 'cbf27d[6](https://github.com/Nordix/eredis/runs/7129339382?check_suite_focus=true#step:7:7)ab[7](https://github.com/Nordix/eredis/runs/7129339382?check_suite_focus=true#step:7:8)621a773539ff4a1c434f33c[8](https://github.com/Nordix/eredis/runs/7129339382?check_suite_focus=true#step:7:9)f3e535'.
You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.
If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:
  git switch -c <new-branch-name>
Or undo this operation with:
  git switch -
Turn off this advice by setting config variable advice.detachedHead to false
escript: exception error: undefined function rebar3:main/1
  in function  escript:run/2 (escript.erl, line 758)
  in call from escript:start/1 (escript.erl, line 277)
  in call from init:start_em/1 
  in call from init:do_boot/3```

zuiderkwast commented:
I have restarted the failing jobs twice, but it fails to install Elvis every time. I don't know why.

Maybe it has something to do with rebar3. The latest rebar3 supports OTP 23-25, so not 22 anymore. Maybe we should remove 22 for eredis too? Or use an older rebar3 for OTP 22?

dhruvjain99 commented:
@zuiderkwast The install elvis is still failing due to this [issue](https://github.com/erlang/rebar3/issues/2025). It is strange to encounter this issue given our versions of rebar, erlang and elvis are locked.

dhruvjain99 commented:
@zuiderkwast I have resolved the issue for OTP 22. The workflows are passing on my [fork](https://github.com/dhruvjain99/eredis/actions) now. It should pass here as well. You may trigger it.

zuiderkwast commented on test/eredis_tcp_SUITE.erl:
```diff
@@ -718,6 +752,22 @@ c() ->
     C.
 
+c_sentinel() ->
+    Res = eredis:start_link([{host, "127.0.0.8"},
+                             {port, ?WRONG_PORT},
+                             {database, 0},
+                             {password, ""},
+                             {reconnect_sleep, 100},
```
When sentinel is used, typically host and port should not be specified, right? I think it's better to remove them here to make it simpler and more realistic. (But let's keep `?WRONG_PORT` in at least one test case like `t_connect_eredis_sentinel` to make sure it is covered too.)

Empty password, db 0, etc. are also redundant and tested elsewhere. The function `c/0` above uses default opts for everything.

zuiderkwast reviewed (commented):
Very good, I think this can be merged soon. I had a second look and just a few minor suggestion.

What about testing crashing sentinels and other failure stuff? Like kill the eredis_sentinel process or kill the conn_pid in a test case? For HA, which is what sentinel is for, I think it's important that these things work as expected. Maybe in a later PR? :)

dhruvjain99 commented:
@zuiderkwast I have added new sentinel related tests and moved them to a separate file. Please let me know if anything else needs to be done for release.

zuiderkwast reviewed (commented):
Very nice tests and very good to have a separate test suite for sentinel!

Only one more thing is some simple documentation. We should describe the sentinel option in the edoc section for `eredis:start_link/1` and mention sentinel support in README.

zuiderkwast commented on test/eredis_sentinel_SUITE.erl:
```diff
@@ -0,0 +1,165 @@
+    process_flag(trap_exit, true),
+    C = c_sentinel(),
+    timer:sleep(100),
+    ?assertMatch(true, is_process_alive(whereis(mymaster))),
```
Btw, to assert that something is true, just use `?assert(is_process_alive(whereis(mymaster)))`

dhruvjain99 replied:
@zuiderkwast We do need sleep as `c_sentinel()` doesn't block until `init/1` is finished. Inside init/1 only we create the sentinel process if not alive.

zuiderkwast replied:
You're right. Actually, only the `eredis_client` process is started. The connection to Redis and even creating the sentinel process is asynchronous, via `self() ! initiate_connection` in `eredis_client:init/1`. So keep the sleep. :-)

dhruvjain99 commented:
@zuiderkwast when can we plan to release this feature?

zuiderkwast commented:
> @zuiderkwast when can we plan to release this feature?

Once it's merged I can tag a new version. I think it can be done within a few days.

zuiderkwast commented on src/eredis_sentinel.erl:
```diff
@@ -0,0 +1,242 @@
+
+-record(eredis_sentinel_state, {
+                                master_group    :: atom(),
+                                addresses       :: [{string(), integer()}],
```
Host can also be `inet:local_address()` (a tuple like `{local, string()}`).

zuiderkwast commented on src/eredis.erl:
```diff
@@ -64,7 +64,12 @@ start_link() ->
+%% <dt>`{sentinel, SentinelOptions}'</dt><dd> Enables client for sentinel mode; options are required
+%% to connect to sentinel cluster; default: undefined </dd>
+%% <dd> Sentinel Options: `{master_group, master_group_name}' - Atom, default: mymaster;
+%% `{addresses, [{Host, Port}]}' - List of {Host, Port} tuples, default: [{"127.0.0.1", 26379}];
```
I'm thinking about the name `addresses` here. An address is an IP address, but here we have a list of `{Host, Port}` tuples. A name like hostportpairs is not very nice though... What do you think about calling it `endpoints`?

(We will not be able to change this in the future. That's why I want to be extra careful with the API including option names.)

zuiderkwast reviewed (approved):
Nice, I'm happy now. :-) I'll merge and make a release tomorrow or Friday.
